### PR TITLE
Use more thread pools in BACKUP/RESTORE to avoid its hanging in tests

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Not for changelog

This PR is a follow-up to https://github.com/ClickHouse/ClickHouse/pull/55132, it adds two more thread pools to handle `BACKUP ON CLUSTER ASYNC` and `RESTORE ON CLUSTER ASYNC` without possible hanging. The PR also introduces lazily creation of such thread pools.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit ca01b5a2d8d0a4f33d4938c5019c2153f483c6ea with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55216/ca01b5a2d8d0a4f33d4938c5019c2153f483c6ea/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6393601885">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55216/ca01b5a2d8d0a4f33d4938c5019c2153f483c6ea/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Backups/BackupsWorker.cpp [hunk 1]
```diff
@@ -218,42 +218,145 @@ namespace
...
+            }
+            if (thread_pool)
+                thread_pool->wait();
+        }
+    }
+
+private:
+    const size_t num_backup_threads;
+    const size_t num_restore_threads;
+    std::unique_ptr<ThreadPool> thread_pools[static_cast<size_t>(ThreadPoolId::MAX)] TSA_GUARDED_BY(mutex);
```
Comment (alesapin) [inline src/Backups/BackupsWorker.cpp:339]: Maybe better to make unordered map for simplicity?

**Review by alesapin — APPROVED** [review]
Too complex and general solution, which doesn't protect from the mentioned error. But ok, let's do it this way.

**Comment by alesapin** [comment]
@vitlibar WDYT?

File: src/Backups/BackupsWorker.cpp (hunk 1 above)
Comment (vitlibar) [reply src/Backups/BackupsWorker.cpp]: done

**Comment by vitlibar** [comment]
> @vitlibar WDYT?

I rewrote my code using `std::map`.

**Comment by vitlibar** [comment]
> Too complex and general solution, which doesn't protect from the mentioned error. But ok, let's do it this way.

Why it doesn't protect? I mean now the parallel tasks waiting each others are going to be always in different thread pools. And the reason of that deadlock was exactly that a task could wait another task from the same thread pool.

**Review by alesapin — APPROVED** [review]

**Pull request merged**